### DISCUSSION TITLE
objective c analysis support in codechecker?

### DISCUSSION THREAD

**Discussion by lipeng28** [Ericsson/codechecker#2799] [Q&A] [answered]
Hi There

Is codechecker supporting analyzing objective c code?  If no do you have the plan to support it?

Thanks
Peng

**Comment by gyorb** [comment, marked as answer]
Hi,

Both the [clang static analyzer](https://clang.llvm.org/docs/analyzer/checkers.html) and [clang-tidy](https://clang.llvm.org/extra/clang-tidy/) can analyze objective c code.
The clang static anlyzer can create plist report files which can be stored by CodeChecker.
For clang-tidy you will have to use the [report-converter](https://github.com/Ericsson/codechecker/tree/master/tools/report-converter) to convert the tidy output into a format which can be stored into a CodeChecker server.

I'm not sure how the current `CodeChecker analyze` part which executes the analyzers would work in an objective c build environment, I'm not familiar with that.

I think a solution which could work now is if you can execute the analyzers on your objective c code (convert the reports if needed), than you can store, view, manage the reports on the web UI or generate html reports.

Br,
Gyorgy

**Comment by lipeng28** [comment]
Hi Gyorgy

Thank you for replying. 
I used the 
CodeChecker log -b "xcodebuild build ..." -o compilation.json
command to dump the json format compilation, however the compilation.json file is empty. 
I guess it may be caused by that intercept-build does not support xcodebuild like compilation.

Best,
Peng

**Comment by gyorb** [comment]
Hi Peng, 

[This](https://sarcasm.github.io/notes/dev/compilation-database.html) is a great collection of tools, which can produce the compile command database.
Based on the description [xcpretty](https://sarcasm.github.io/notes/dev/compilation-database.html#xcpretty) can produce a compile command database but I did not test it. Maybe it is worth a try.

Br,
Gyorgy

**Comment by lipeng28** [comment]
Thanks for pointing out xcpretty, I have found it and integrated it in CodeChecker, it works to generate json file successfully.
Best,
Peng

**Comment by whisperity** [comment]
@lipeng28 Once your compilation database is ready, can you run the analysis itself to completion?

You should look at the checker list, and perhaps will need to enable checkers from the `osx` group, to ensure issues modelled by Clang for Darwin code is found. As far as I know, codechecker doesn't enable those automatically.

**Comment by lipeng28** [comment]
Hi Whisperity

No, I can not run the analysis to completion, I have extended the implementation to finish it. And thanks for pointing out osx group is not enabled by default.

Best,
Peng

**Comment by whisperity** [comment]
You can see here which checkers are enabled by default and in additional profiles: https://github.com/Ericsson/codechecker/blob/master/config/config.json#L18 You can say `CodeChecker checkers` to list the available checkers, and to `analyze`, you can say `-enable osx` instead of having to manually specify *every* OSX-related checker.

**Comment by whisperity** [comment]
You can also add the `--verbose debug_analyzer` flag to `analyze` to ask CodeChecker to dump the native analyser binary calls. CodeChecker analyses your compilation database and creates invocations, but the analysis itself is done by ClangSA and Tidy, which are open-source projects and there are people and companies behind them who use OS/X. So by checking the native analyser calls and trying them out individually, we can see if the issue comes from CodeChecker or from Clang itself.
